src/replication.c

```diff
@@ -497,292 +180,72 @@ perform_read_write_on_fd(int fd, uint8_t *data, uint64_t len, int *errorno,
-	rcommon_cmd_t *rcommq_ptr = NULL;
-	zvol_io_hdr_t *io_rsp_hdr = replica->io_resp_hdr;
-	void *data = replica->io_resp_data;
-	uint64_t datalen = 0;
-	uint8_t *dataptr = NULL;
-
-	MTX_LOCK(&replica->r_mtx);
-	//Find IO in read queue, signal luworker, and dequeue
-	TAILQ_FOREACH(rep_cmd, &replica->read_waitq, rread_cmd_next) {
-		if(io_rsp_hdr->io_seq == rep_cmd->io_seq) {
-			io_found = true;
-			rep_cmd->ack_recvd = true;
-			TAILQ_REMOVE(&replica->read_waitq, rep_cmd, rread_cmd_next);
-			MTX_UNLOCK(&replica->r_mtx);
-			if (io_rsp_hdr->status == ZVOL_OP_STATUS_OK)
-				dataptr = get_read_resp_data(io_rsp_hdr, data, &datalen);
-			MTX_LOCK(&spec->rcommonq_mtx);
-			rcommq_ptr = rep_cmd->rcommq_ptr;
-			MTX_LOCK(&rcommq_ptr->rcommand_mtx);
-			if (io_rsp_hdr->status == ZVOL_OP_STATUS_OK)
-				rcommq_ptr->status = 1;
-			else
-				rcommq_ptr->status = -1;
-			rcommq_ptr->data = dataptr;
-			rcommq_ptr->data_len = datalen;
-			luworker_id = rcommq_ptr->luworker_id;
-			put_to_mempool(&rcmd_mempool, rep_cmd);
-
-			signal_luworker();
-
-			TAILQ_REMOVE(&spec->rcommon_waitq, rcommq_ptr, wait_cmd_next);
-			rcommq_ptr->state = CMD_EXECUTION_DONE;
-			if(rcommq_ptr->completed) {
-				MTX_UNLOCK(&rcommq_ptr->rcommand_mtx);
-				MTX_UNLOCK(&spec->rcommonq_mtx);
-				clear_rcomm_cmd(rcommq_ptr);
-			} else {
-				rcommq_ptr->completed = true;
-				MTX_UNLOCK(&rcommq_ptr->rcommand_mtx);
-				MTX_UNLOCK(&spec->rcommonq_mtx);
-			}
-			free(data);
-			return 0;
-		}
-	}
-	MTX_UNLOCK(&replica->r_mtx);
-	if(!io_found) {
-		/* a print or a stats */
-		return -1;
-	}
-	return 0;
-}
+	uint8_t *dataptr = resp->data_ptr;
+	uint64_t len = 0, parsed = 0;
+	io_data_chunk_t  *io_chunk_entry;
 
-//else case for matching with copies_sent will be handled outside handle_consistency_met
-#define handle_consistency_met() {\
-	rcommq_ptr->status = 1; \
-	TAILQ_REMOVE(&spec->rcommon_waitq, rcommq_ptr, wait_cmd_next); \
-	if ((rcommq_ptr->acks_recvd + rcommq_ptr->ios_aborted) != rcommq_ptr->copies_sent) { \
-		rcommq_ptr->state = CMD_ENQUEUED_TO_PENDINGQ; \
-		TAILQ_INSERT_TAIL(&spec->rcommon_pendingq, rcommq_ptr, pending_cmd_next); \
-	} \
-	signal_luworker(); \
-}
+	if (!TAILQ_EMPTY(io_chunk_list)) {
+		io_chunk_entry = TAILQ_FIRST(io_chunk_list);
 
-#define handle_all_resp_recvd() { \
-	if(rcommq_ptr->state == CMD_ENQUEUED_TO_PENDINGQ) { \
-		TAILQ_REMOVE(&spec->rcommon_pendingq, rcommq_ptr, pending_cmd_next); \
-	} else if(rcommq_ptr->state == CMD_ENQUEUED_TO_WAITQ) { \
-		TAILQ_REMOVE(&spec->rcommon_waitq, rcommq_ptr, wait_cmd_next); \
-	} \
-	rcommq_ptr->state = CMD_EXECUTION_DONE; \
-	if(rcommq_ptr->completed) { \
-		MTX_UNLOCK(&rcommq_ptr->rcommand_mtx); \
-		clear_rcomm_cmd(rcommq_ptr); \
-		rcommq_ptr = NULL; \
-	} else { \
-		rcommq_ptr->completed = true; \
-		if (rcommq_ptr->acks_recvd < spec->consistency_factor) { \
-			rcommq_ptr->status = -1; \
-			signal_luworker(); \
-		} \
-		MTX_UNLOCK(&rcommq_ptr->rcommand_mtx); \
-	} \
-}
+		while (parsed < hdr->len) {
```

Comment:
there is data_len in the replica_rcommon_resp_t.. not sure whether it is in use (yet to go through complete PR).. if not in use, it can be removed